server/neptune/gateway/server.go

```diff
@@ -266,82 +269,101 @@ func NewServer(config Config) (*Server, error) {
 		pullConverter,
 		vcsClient,
 		featureAllocator,
+		asyncScheduler,
 	)
 
 	router := mux.NewRouter()
+	router.HandleFunc("/healthz", Healthz).Methods("GET")
+	router.HandleFunc("/status", statusController.Get).Methods("GET")
+	router.HandleFunc("/events", gatewayEventsController.Post).Methods("POST")
+
+	n := negroni.New(&negroni.Recovery{
+		Logger:     log.New(os.Stdout, "", log.LstdFlags),
+		PrintStack: false,
+		StackAll:   false,
+		StackSize:  1024 * 8,
+	})
+	n.UseHandler(router)
+
+	s := httpInternal.ServerProxy{
+		Server: &http.Server{
+			Addr:    fmt.Sprintf(":%d", config.Port),
+			Handler: n,
+		},
+		SSLCertFile: config.SSLCertFile,
+		SSLKeyFile:  config.SSLKeyFile,
+	}
 
 	return &Server{
-		StatsCloser:      closer,
-		Router:           router,
-		EventController:  gatewayEventsController,
-		StatusController: statusController,
-		Logger:           ctxLogger,
-		Port:             config.Port,
-		Drainer:          drainer,
+		StatsCloser: closer,
+		Handler:     n,
+		Scheduler:   asyncScheduler,
+		Logger:      ctxLogger,
+		Port:        config.Port,
+		Drainer:     drainer,
+		Server:      s,
 	}, nil
 
 }
 
 // Start is blocking and listens for incoming requests until a configured shutdown
 // signal is received.
 func (s *Server) Start() error {
-
-	// TODO: remove router initialization from here
-	// I assume this is currently happening to ensure that healthz is returning ready
-	// only when we are actually ready to receive requests, however, a better to do this is
-	// by using some atomic value to determine when our server is ready to start taking traffic.
-	// This way we'd have clean separation between setup and execution
-	s.Router.HandleFunc("/healthz", s.Healthz).Methods("GET")
-	s.Router.HandleFunc("/status", s.StatusController.Get).Methods("GET")
-	s.Router.HandleFunc("/events", s.EventController.Post).Methods("POST")
-
-	n := negroni.New(&negroni.Recovery{
-		Logger:     log.New(os.Stdout, "", log.LstdFlags),
-		PrintStack: false,
-		StackAll:   false,
-		StackSize:  1024 * 8,
-	})
-	n.UseHandler(s.Router)
-
 	defer s.Logger.Close()
 
-	// Ensure server gracefully drains connections when stopped.
-	stop := make(chan os.Signal, 1)
-	// Stop on SIGINTs and SIGTERMs.
-	signal.Notify(stop, os.Interrupt, syscall.SIGTERM)
+	// we create a base context that is marked done when we get a sigterm.
+	// we should use this context for other async work to ensure we
+	// are gracefully handling shutdown and not dropping data.
+	mainCtx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
+	defer stop()
 
-	server := &http.Server{Addr: fmt.Sprintf(":%d", s.Port), Handler: n}
-	go func() {
+	// error group here makes it easier to add other processes and share a ctx between them
+	group, gCtx := errgroup.WithContext(mainCtx)
+	group.Go(func() error {
```

Comment:
TIL, neat!

Reply:
yeah we should probably use this for the temporal worker.